authorization/src/main/kotlin/org/modelix/authorization/permissions/SchemaInstance.kt

```diff
@@ -0,0 +1,180 @@
+        }
+
+        updateIncludes()
+        return newInstance
+    }
+
+    private fun evaluateExpression(expr: IExpression, source: ResourceInstance): String {
+        return when (expr) {
+            is AddPrefix -> expr.prefix + evaluateExpression(expr.expr, source)
+            is SourceParameterValue -> source.reference.parameterValues[source.resourceSchema.parameters.indexOf(expr.name)]
+        }
+    }
+
+    fun instantiatePermission(ref: PermissionInstanceReference): ResourceInstance.PermissionInstance {
+        return instantiateResource(ref.resource)
+            .getOrCreatePermissionInstance(ref.permissionName)
+    }
+
+    fun updateIncludes() {
+        resources.values.forEach { it.updateIncludes() }
+    }
+
+    inner class ResourceInstance(
+        val parent: ResourceInstance?,
+        val resourceSchema: Resource,
+        val reference: ResourceInstanceReference,
+    ) {
+        val childResources: MutableMap<ResourceInstanceReference, ResourceInstance> = LinkedHashMap()
+        val relationTargetInstances: MutableMap<String, ResourceInstance> = LinkedHashMap()
+        val permissions: MutableMap<String, PermissionInstance> = LinkedHashMap()
+
+        fun getOrCreatePermissionInstance(name: String): PermissionInstance {
+            return permissions.getOrPut(name) {
+                val permissionSchema = requireNotNull(resourceSchema.permissions[name]) {
+                    "Permission '$name' not found in $reference"
+                }
+                PermissionInstance(permissionSchema, PermissionInstanceReference(name, reference))
+            }
+        }
+
+        fun updateIncludes() {
+            childResources.values.forEach { it.updateIncludes() }
+            permissions.values.forEach { it.updateIncludes() }
+        }
+
+        fun resolveResourceInstance(name: String): ResourceInstance? {
+            if (resourceSchema.name == name) return this
+
+            childResources.values.firstOrNull { it.resourceSchema.name == name && it.resourceSchema.parameters.isEmpty() }?.let { return it }
+            if (parent != null) {
+                parent.resolveResourceInstance(name)
+            } else {
+                resources.values.firstOrNull { it.resourceSchema.name == name && it.resourceSchema.parameters.isEmpty() }
+            }?.let { return it }
+
+            relationTargetInstances.values.asSequence().map { it.resolveResourceInstance(name) }.firstOrNull()?.let { return it }
+
+            return null
+        }
+
+        override fun toString() = reference.toString()
+
+        inner class PermissionInstance(val permissionSchema: Permission, val ref: PermissionInstanceReference) {
+            val includedIn: MutableSet<PermissionInstance> = HashSet()
+            val includes: MutableSet<PermissionInstance> = HashSet()
+
+            fun transitiveIncludes(acc: MutableSet<PermissionInstance> = LinkedHashSet()): Set<PermissionInstance> {
```

Comment:
`transitiveIncludes` and `transitiveIncludedIn` seems unused

Reply:
Removed.